### PR TITLE
373 add code for spade vae gan

### PR DESCRIPTION
**Pull request by virginiafdez**
- Added support for SPADE-GAN network (SPADE normalisation and GAN-VAEGAN-like generator)
- Added unit tests
- Added tutorial for 2D usage of SPADE-GAN along with Multi-Scale Discriminator.

### CODE REVIEW

File: generative/networks/nets/spade_network.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+from monai.utils.enums import StrEnum
+import torch.nn.functional as F
+from generative.losses.kld_loss import KLDLoss
+
+class UpsamplingModes(StrEnum):
+    bicubic = "bicubic"
+    nearest = "nearest"
+    bilinear = "bilinear"
+
+class SPADE_ResNetBlock(nn.Module):
```
Comment (marksgraham) [inline generative/networks/nets/spade_network.py:28]: please add a docstring

File: generative/networks/nets/spade_network.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        return out
+
+    def shortcut(self, x, seg):
+        if self.learned_shortcut:
+            x_s = self.conv_s(self.norm_s(x, seg))
+        else:
+            x_s = x
+        return x_s
+
+class SPADE_Encoder(nn.Module):
```
Comment (marksgraham) [inline generative/networks/nets/spade_network.py:91]: docstrng here too

File: generative/networks/nets/spade_network.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                                      act = None,
+                                      norm = None,
+                                      kernel_size=1,
+                                      )
+            self.norm_s = SPADE(label_nc=label_nc, norm_nc=self.in_channels, kernel_size=kernel_size,
+                                spatial_dims=spatial_dims, hidden_channels=spade_intermediate_channels,
+                                norm=norm)
+
+    def forward(self, x, seg):
+
```
Comment (marksgraham) [inline generative/networks/nets/spade_network.py:77]: remove blank lines between the def and the function content

File: generative/networks/nets/spade_network.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+                                     padding=(kernel_size-1)//2,
+                                     kernel_size=kernel_size,
+                                     norm = None,
+                                     act=last_act
+                                     )
+
+
+    def forward(self, seg, z: torch.Tensor = None):
+
+
```
Comment (marksgraham) [inline generative/networks/nets/spade_network.py:218]: remove blank lines

**Review by marksgraham — CHANGES_REQUESTED** [review]
Hi,

Looks good! Some changes requested in the comments

File: tutorials/generative/2d_autoencoderkl/2d_autoencoderkl_tutorial.ipynb [hunk 5]
```diff
@@ -115,7 +116,7 @@
     "import shutil\n",
     "import tempfile\n",
     "import time\n",
-    "\n",
+    "from pathlib import Path\n",
```
[the remark this answers is not in the document]
Comment (virginiafdez) [reply tutorials/generative/2d_autoencoderkl/2d_autoencoderkl_tutorial.ipynb:119]: I see this line in the tutorial, a bit above

File: generative/networks/blocks/spade_norm.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+        # Part 1. generate parameter-free normalized activations
+        normalized = self.param_free_norm(x)
+
+        # Part 2. produce scaling and bias conditioned on semantic map
+        segmap = F.interpolate(segmap, size=x.size()[2:], mode='nearest')
+        actv = self.mlp_shared(segmap)
+        gamma = self.mlp_gamma(actv)
+        beta = self.mlp_beta(actv)
+        out = normalized * (1 + gamma) + beta
```
[the remark this answers is not in the document]
Comment (virginiafdez) [reply generative/networks/blocks/spade_norm.py:79]: It's in their original code

[the remark this answers is not in the document]
Comment (virginiafdez) [reply generative/networks/blocks/spade_norm.py:79]: It comes out (I think) of taking h out of equation (1) (replacing mu by the content of equation 2)

**Comment by virginiafdez** [comment]
Added requested changes:

- Added more info to the tutorial and loss plot
- Removed blank lines
- Added docstrings

**Review by marksgraham — APPROVED** [review]
